diff --git a/defs.h b/defs.h
index 82fb982..c9d0f89 100644
--- a/defs.h
+++ b/defs.h
@@ -33,6 +33,7 @@ void            fileinit(void);
 int             fileread(struct file*, char*, int n);
 int             filestat(struct file*, struct stat*);
 int             filewrite(struct file*, char*, int n);
+void						slab_add_page(void);
 
 // fs.c
 void            readsb(int dev, struct superblock *sb);
diff --git a/file.c b/file.c
index 24b32c2..90df708 100644
--- a/file.c
+++ b/file.c
@@ -9,17 +9,21 @@
 #include "spinlock.h"
 #include "sleeplock.h"
 #include "file.h"
+#include "mmu.h"
 
 struct devsw devsw[NDEV];
-struct {
-  struct spinlock lock;
-  struct file file[NFILE];
-} ftable;
+
+static struct {
+	struct spinlock lock;
+	void *freelist;
+} struct_file_cache;
 
 void
 fileinit(void)
 {
-  initlock(&ftable.lock, "ftable");
+  initlock(&struct_file_cache.lock, "structfilecache");
+	struct_file_cache.freelist = 0;
+	slab_add_page();
 }
 
 // Allocate a file structure.
@@ -28,15 +32,19 @@ filealloc(void)
 {
   struct file *f;
 
-  acquire(&ftable.lock);
-  for(f = ftable.file; f < ftable.file + NFILE; f++){
-    if(f->ref == 0){
-      f->ref = 1;
-      release(&ftable.lock);
-      return f;
-    }
+  acquire(&struct_file_cache.lock);
+  if(!struct_file_cache.freelist)
+    slab_add_page();
+  f = struct_file_cache.freelist;
+  if(f) {
+    struct_file_cache.freelist = f->next;
+    f->next = 0;
+    f->ref = 1;
+    release(&struct_file_cache.lock);
+    return f;
   }
-  release(&ftable.lock);
+  release(&struct_file_cache.lock);
+
   return 0;
 }
 
@@ -44,11 +52,11 @@ filealloc(void)
 struct file*
 filedup(struct file *f)
 {
-  acquire(&ftable.lock);
+  acquire(&struct_file_cache.lock);
   if(f->ref < 1)
     panic("filedup");
   f->ref++;
-  release(&ftable.lock);
+  release(&struct_file_cache.lock);
   return f;
 }
 
@@ -58,17 +66,19 @@ fileclose(struct file *f)
 {
   struct file ff;
 
-  acquire(&ftable.lock);
+  acquire(&struct_file_cache.lock);
   if(f->ref < 1)
     panic("fileclose");
   if(--f->ref > 0){
-    release(&ftable.lock);
+    release(&struct_file_cache.lock);
     return;
   }
   ff = *f;
   f->ref = 0;
   f->type = FD_NONE;
-  release(&ftable.lock);
+  f->next = struct_file_cache.freelist;
+  struct_file_cache.freelist = f;
+  release(&struct_file_cache.lock);
 
   if(ff.type == FD_PIPE)
     pipeclose(ff.pipe, ff.writable);
@@ -155,3 +165,16 @@ filewrite(struct file *f, char *addr, int n)
   panic("filewrite");
 }
 
+void
+slab_add_page(void)
+{
+	char *page = kalloc();
+	if(!page)
+		return;
+	int n = PGSIZE / sizeof(struct file);
+	for(int i = 0; i < n; i++) {
+		struct file *f = (struct file*)page + i;
+		f->next = struct_file_cache.freelist;
+		struct_file_cache.freelist = f;
+	}
+}
diff --git a/file.h b/file.h
index 0990c82..ea8cb61 100644
--- a/file.h
+++ b/file.h
@@ -6,6 +6,7 @@ struct file {
   struct pipe *pipe;
   struct inode *ip;
   uint off;
+  struct file *next;
 };
 
 
diff --git a/param.h b/param.h
index a7e90ef..eb65808 100644
--- a/param.h
+++ b/param.h
@@ -2,7 +2,6 @@
 #define KSTACKSIZE 4096  // size of per-process kernel stack
 #define NCPU          8  // maximum number of CPUs
 #define NOFILE       16  // open files per process
-#define NFILE       100  // open files per system
 #define NINODE       50  // maximum number of active i-nodes
 #define NDEV         10  // maximum major device number
 #define ROOTDEV       1  // device number of file system root disk
